src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/InstructionList.cs

```diff
@@ -407,13 +407,10 @@ internal void SwitchToBoxed(int index, int instructionIndex)
         {
             var instruction = _instructions[instructionIndex] as IBoxableInstruction;
 
-            if (instruction != null)
+            Instruction newInstruction = instruction?.BoxIfIndexMatches(index);
```

Comment:
It does, and I'm not sure about the optimisation from the compiler. It's definitely not an obvious tidy-up and I'm aiming to have this PR just changes that are obviously correct rather than anything clever or where some gains outweigh some loses, so I'll revert this change.